fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/temp/KeyExpressionComparisons.java

```diff
@@ -0,0 +1,270 @@
+                                            @Nonnull List<KeyExpressionWithComparison> children,
+                                            @Nonnull ComparisonRange comparison) {
+            this.root = root;
+            this.keyExpression = keyExpression;
+            this.children = children;
+            this.comparison = comparison;
+        }
+
+        public MatchedComparisonType getMatchedComparisonType() {
+            if (keyExpression instanceof KeyExpressionWithoutChildren) {
+                return MatchedComparisonType.from(comparison);
+            }
+            if (keyExpression instanceof ThenKeyExpression) {
+                boolean seenInequality = false;
+                for (KeyExpressionWithComparison child : children) {
+                    switch (child.getMatchedComparisonType()) {
+                        case NOT_MATCHED:
+                            return MatchedComparisonType.NOT_MATCHED;
+                        case INEQUALITY_MATCHED:
+                            seenInequality = true;
+                            break;
+                        default:
+                            continue;
+                    }
+                }
+                if (seenInequality) {
+                    return MatchedComparisonType.INEQUALITY_MATCHED;
+                } else {
+                    return MatchedComparisonType.MATCHED;
+                }
+            }
+            // TODO
+            return MatchedComparisonType.NOT_MATCHED;
+        }
+
+        @Nullable
+        public ComparisonRange getComparison() {
+            return comparison;
+        }
+
+        @Nonnull
+        private KeyExpressionWithComparison withComparison(@Nonnull ComparisonRange comparison) {
+            return new KeyExpressionWithComparison(root, keyExpression, children, comparison);
+        }
+
+        private KeyExpressionWithComparison withChild(@Nonnull KeyExpressionWithComparison newChild) {
+            return withChildren(Collections.singletonList(newChild));
+        }
+
+        @Nonnull
+        private KeyExpressionWithComparison withChildren(@Nonnull List<KeyExpressionWithComparison> newChildren) {
+            return new KeyExpressionWithComparison(root, keyExpression, newChildren, comparison);
+        }
+
+        public Optional<KeyExpressionWithComparison> matchWith(@Nonnull ComponentWithComparison component) {
+            if (keyExpression instanceof ThenKeyExpression) {
+                return matchWithThen(component);
+            }
+            if (keyExpression instanceof GroupingKeyExpression) {
+                return children.get(0).matchWith(component).map(this::withChild);
+            }
+            if (keyExpression instanceof KeyWithValueExpression) {
+                return children.get(0).matchWith(component).map(this::withChild);
+            }
+            if (keyExpression instanceof FieldKeyExpression && component instanceof FieldWithComparison) {
```

Comment:
Yes, exactly. That will come in #367 because implementing it properly requires more advanced data structure manipulations than are included in this PR.